persistence/sql/identity/persister_identity.go

```diff
@@ -181,82 +188,114 @@ func (p *IdentityPersister) FindByCredentialsIdentifier(ctx context.Context, ct
 	return i.CopyWithoutCredentials(), creds, nil
 }
 
+var credentialsTypes = map[identity.CredentialsType]*identity.CredentialsTypeTable{}
+
 func (p *IdentityPersister) findIdentityCredentialsType(ctx context.Context, ct identity.CredentialsType) (_ *identity.CredentialsTypeTable, err error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.findIdentityCredentialsType")
 	defer otelx.End(span, &err)
 
+	if v, ok := credentialsTypes[ct]; ok && v != nil {
+		return v, nil
+	}
+
 	var m identity.CredentialsTypeTable
 	if err := p.GetConnection(ctx).Where("name = ?", ct).First(&m); err != nil {
 		return nil, sqlcon.HandleError(err)
 	}
+	credentialsTypes[ct] = &m
 	return &m, nil
 }
 
-func (p *IdentityPersister) createIdentityCredentials(ctx context.Context, i *identity.Identity) (err error) {
+func (p *IdentityPersister) createIdentityCredentials(ctx context.Context, conn *pop.Connection, identities ...*identity.Identity) (err error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.createIdentityCredentials")
 	defer otelx.End(span, &err)
 
-	c := p.GetConnection(ctx)
+	var (
+		nid         = p.NetworkID(ctx)
+		traceConn   = &batch.TracerConnection{Tracer: p.r.Tracer(ctx), Connection: conn}
+		credentials []*identity.Credentials
+		identifiers []*identity.CredentialIdentifier
+	)
 
-	nid := p.NetworkID(ctx)
-	for k := range i.Credentials {
-		cred := i.Credentials[k]
+	for _, ident := range identities {
```

Comment:
This may be buggy because your data structure is a slice with pointers. See https://medium.com/swlh/use-pointer-of-for-range-loop-variable-in-go-3d3481f7ffc9

Please use the index to access the variable like you did with `ident.Credentials`